### PR TITLE
Initial Sigstore bundle support

### PR DESCRIPTION
**Pull request by woodruffw**
This adds initial support for Sigstore bundles during signing, in the form of the `--bundle` flag. When passed, `sigstore sign` will generate a single `{input}.sigstore` instead of separate `.crt`, `.sig`, and `.rekor` files.

I haven't included `sigstore verify` support in the initial changeset, in an effort to keep the diff small. But adding it shouldn't be difficult, and will provide a good dogfood/smoke test for round-tripping through the bundle format.

See #251.

Signed-off-by: William Woodruff <william@trailofbits.com>

### CODE REVIEW

**Comment by woodruffw** [comment]
Example generated bundle:

```json
{
    "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.1",
    "verificationMaterial": {
        "x509CertificateChain": {
            "certificates": [
                {
                    "rawBytes": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN3akNDQWtlZ0F3SUJBZ0lVU0NpWUVuTzVLa0g0YWxaNTlDNmEyMUVwUENZd0NnWUlLb1pJemowRUF3TXcKTnpFVk1CTUdBMVVFQ2hNTWMybG5jM1J2Y21VdVpHVjJNUjR3SEFZRFZRUURFeFZ6YVdkemRHOXlaUzFwYm5SbApjbTFsWkdsaGRHVXdIaGNOTWpNd01USXdNVFkxTkRBeVdoY05Nak13TVRJd01UY3dOREF5V2pBQU1IWXdFQVlICktvWkl6ajBDQVFZRks0RUVBQ0lEWWdBRW5RRUZsMmUzdzdySjQ5eHdGbWVheWN3ZnZTajBMWXB1WFZ1d3R0aE4KaTF3TWY0ZGF0SGNSeFpMN3ZTaXp1VHY5M2twZEZ5UTBhOGJYeTBTYUpia2NBUFVEVDFTY05VVkZwOEVOOTdTTApBbmJ2dzJLRlk1YjF4QmtwR0VwajMwNmVvNElCU1RDQ0FVVXdEZ1lEVlIwUEFRSC9CQVFEQWdlQU1CTUdBMVVkCkpRUU1NQW9HQ0NzR0FRVUZCd01ETUIwR0ExVWREZ1FXQkJUdFpicTBIYmk5RWc3eThIaXZQKzFvV2NNNzh6QWYKQmdOVkhTTUVHREFXZ0JSeGhqQ21GSHhpYi9uMzF2UUZHbjlmLyt0dnJEQWpCZ05WSFJFQkFmOEVHVEFYZ1JWMwphV3hzYVdGdFFIbHZjM05oY21saGJpNXVaWFF3TEFZS0t3WUJCQUdEdnpBQkFRUWVhSFIwY0hNNkx5OW5hWFJvCmRXSXVZMjl0TDJ4dloybHVMMjloZFhSb01JR0tCZ29yQmdFRUFkWjVBZ1FDQkh3RWVnQjRBSFlBS3pDODNHaUkKeWVMaDJDWXBYblFmU0RreGxnTHluRFBMWGtOQS9yS3Nobm9BQUFHRjBCbitCZ0FBQkFNQVJ6QkZBaUJ6TENLbwpOUzlqS01iaHYwU2FteGdoWUJQK3RlbExGRzlYcDBBVXlQQmlpUUloQUxXTXpzU1pBaVdYZHN6S0xXQU15b01LClBNcjVOWndTM1Q4SmE0ZFhWbUJWTUFvR0NDcUdTTTQ5QkFNREEya0FNR1lDTVFDNGU1cktGcWRHbEZkTzJEemkKR3k3ajZJVkpZdENhQStraitUL0MwVWhwZ0xON2h2dGVXOXdkSUwrU1l0bjRvVG9DTVFDMmswNDVFdktBcTdHRwpEbngwaEhrTWJmS2ZKK0NCZjV1bjZSODkxOWpST1ZVdUdEeUlLUWhFSi9RQjBhaVV3bDg9Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K"
                }
            ]
        },
        "tlogEntries": [
            {
                "logIndex": "2242969",
                "logId": {
                    "keyId": "0y8wo8MtY5wrdiIFohx7sHeI5oKDpK5vQhGHI6G+pJY="
                },
                "kindVersion": {
                    "kind": "hashedrekord",
                    "version": "0.0.1"
                },
                "integratedTime": "1674233642",
                "inclusionPromise": {
                    "signedEntryTimestamp": "MEUCIQD31VsczBNtApPFNW31nn20hzxULvVCtLTe7BnAjJCQiwIgN4RlYRkfRSdC7A8E6FufttvH53L5GCE+gmuh2TKVx+I="
                },
                "inclusionProof": {
                    "logIndex": "2228150",
                    "rootHash": "sxhOFoisbFXhJmKOSlimQWBpgkrDT5eCaSaUcydRGEg=",
                    "treeSize": "2228151",
                    "hashes": [
                        "TwqHCU5ELTQy/KKeu+0+Qlnjr0zzPziKXjE5TGXeRT0=",
                        "HtGiDzArXW02VJtZKJVYEIW1cyB7jR0d9c60io54x8s=",
                        "Wjgnj5IgRwHKGVnpu6DKjZWP2ZDRyIvXC2OKO437mQ8=",
                        "ZXWmnfXIqBNf4a2+MosY2T0zeCGACcqlYxrYXdY5INU=",
                        "rkq9/Kqf8fYE4vhLb5vzhhbcsI5kT2JNY06JZmLMlC0=",
                        "O8GhxttbaSwc3hOClnMY2tliEqlpXiUKhuUKPbPYZbA=",
                        "UaBSu2yb0gg+mg7gWVuQViCWJroP3wAo9sm+70DtT94=",
                        "hbzBh170Y7wu6HE4gFaJX1m3tcnffbIbneNVNitxEQ8=",
                        "7RquioMcFFcBwp8L9WgAYvNdmFIX2PLFRO3Ig3I4aG4=",
                        "sEdlyfBp6l2RMcB1BIr7PJyI8d6jWYNBbz4EO0ocBH0=",
                        "PbBHHJ1PoLFq1tjP/Z4nE7yCxtIxUhtqkSdpbkLmrE4=",
                        "d8Hx9AnJ483YnmugSNyUJyzV+dWJfK4wEcn9d9Q+4hM=",
                        "KCciPEwSUhXM04e8244YynDhwh+722/pabo6ZAtt0fc=",
                        "cfT2nw0gal6i8QktNkaFsq6w3Aeu62pnB6sDRzKkAZw=",
                        "VwBj5hN1tw74kRJeHAQaqdSWrXWk7Zb4c1PJfrpiKNw="
                    ]
                },
                "canonicalizedBody": "eyJib2R5IjoiZXlKaGNHbFdaWEp6YVc5dUlqb2lNQzR3TGpFaUxDSnJhVzVrSWpvaWFHRnphR1ZrY21WcmIzSmtJaXdpYzNCbFl5STZleUprWVhSaElqcDdJbWhoYzJnaU9uc2lZV3huYjNKcGRHaHRJam9pYzJoaE1qVTJJaXdpZG1Gc2RXVWlPaUl5TVdKbU9XRmxNR1F5WkRsaFpETmlOMlppTmpSbU1tUTJNV0prTWpFeU1XSXhZemMxWldWaU1UaGlNekJqWmpjeVptRm1NVEV3TW1aa01tRm1ZMlkzSW4xOUxDSnphV2R1WVhSMWNtVWlPbnNpWTI5dWRHVnVkQ0k2SWsxSFdVTk5VVU5IZEdnck1tWjRlRVkxZWpWNVIwVTVaSEZrVVhwWFFXTTJWMFoxVUhCV1ExcEljakF6UTJSM2F6UXpaV2RRVmpWRU9GaE9Wak51WjA5dlJqUkdWMWx2UTAxUlEwd3hkVFZTWVhGTUwwUXlSMGw0ZUdOQ2JHTkVjREJMZEhVcllUQTVhVlZMYlRCUlFVZDFVR1pLTUVZeGVraDFRMmxTVWxndk0zWXdLMWcxVTNOcFNWVTlJaXdpY0hWaWJHbGpTMlY1SWpwN0ltTnZiblJsYm5RaU9pSk1VekIwVEZNeFExSlZaRXBVYVVKRVVsWktWVk5WV2twUk1FWlZVbE13ZEV4VE1IUkRhekZLVTFWT00yRnJUa1JSVjNSc1dqQkdNMU5WU2tKYU1HeFdWVEJPY0ZkVlZuVlVlbFpNWVRCbk1GbFhlR0ZPVkd4RVRtMUZlVTFWVm5kVlJVNWFaREJPYmxkVmJFeGlNWEJLWlcxdmQxSlZSak5VV0dOTFZHNXdSbFpyTVVOVVZXUkNUVlpXUmxFeWFFNVVWMDE1WWtjMWFrMHhTakpaTWpGV1pGWndTRlpxU2s1VmFsSXpVMFZHV2xKR1dsSlZWVkpHWlVaYU5sbFdaR3RsYlZKSVQxaHNZVlY2Um5kWmJUVlRZa0Z3YW1KVVJuTlhhMlJ6WVVkU1NGWllaRWxoUjA1UFZGZHdUbVF3TVZWVFdHUk9Wa1pyZUZSclVrSmxWbVJ2V1RBMVRtRnJNVE5VVmxKS1pEQXhWVmt6WkU5U1JVWTFWakp3UWxGVk1VbFhXR1JHVVZac1NVTnJkSFpYYTJ3MllXcENSRkZXUmxwU2EzTXdVbFZXUWxFd2JFVlhWMlJDVWxjMVVsSlZXbk5OYlZWNlpIcGtlVk5xVVRWbFNHUkhZbGRXYUdWWFRqTmFibHBVWVdwQ1RWZFlRakZYUmxveFpETlNNR0ZGTkV0aFZFWXpWRmRaTUZwSFJqQlRSMDVUWlVad1RVNHpXbFJoV0hBeFZraFpOVTB5ZEhkYVJWbzFWVlJDYUU5SFNsbGxWRUpVV1ZWd2FXRXlUa0pWUmxaRlZrUkdWRmt3TlZaV2ExcDNUMFZXVDA5VVpGUlVRWEJDWW0xS01tUjZTa3hTYkdzeFdXcEdORkZ0ZEhkU01GWjNZV3BOZDA1dFZuWk9SV3hEVlRGU1JGRXdSbFpXV0dSRldqRnNSVlpzU1hkVlJVWlNVME01UTFGV1JrVlJWMlJzVVZVeFExUlZaRUpOVmxaclEydHdVbFZWTVU1UlZ6bElVVEJPZWxJd1JsSldWVnBEWkRBeFJWUlZTWGRTTUVWNFZsZFNSVm94UmxoUmEwcFZaRVp3YVdOVVFrbFpiV3MxVWxkak0yVlVhRWxoV0ZwUlMzcEdkbFl5VGs1T2VtZzJVVmRaUzFGdFpFOVdhMmhVVkZWV1NGSkZSbGhhTUVwVFpVZG9jVkV5TVVkVFNHaHdXV2s1ZFUxNlJqSlZWVnBJWW1wc2JVeDVkREJrYmtwRlVWZHdRMW93TlZkVFJrcEdVV3RHYlU5RlZraFdSVVpaV2pGS1YwMTNjR2hXTTJoNldWWmtSMlJHUmtsaVNGcHFUVEExYjFreU1YTmhSMHB3VGxoV1lWZEdSak5VUlVaYVV6QjBNMWRWU2tOUlZXUkZaRzV3UWxGclJsSlZWMVpvVTBaSmQxa3dhRTVPYTNnMVQxYzFhRmRHU25aRGJWSllVMWhXV2sxcWJEQlVSRW8wWkd4dmVXSklWazFOYW14dldrWm9VMkl3TVVwU01IUkRXakk1ZVZGdFpFWlNWVVpyVjJwV1Fsb3hSa1JSYTJnelVsZFdibEZxVWtKVFJteENVek53UkU5RVRraGhWV3RMWlZkV1RXRkVTa1JYV0VKWllteEdiVlV3VW5KbFIzaHVWRWhzZFZKR1FrMVhSM1JQVVZNNWVWTXpUbTlpYlRsQ1VWVkdTRkpxUWtOaWFYUkRXakJHUWxGclJrNVJWa28yVVd0YVFtRlZTalpVUlU1TVluZHdUMVY2YkhGVE1ERnBZVWhaZDFVeVJuUmxSMlJ2VjFWS1VVc3pVbXhpUlhoSFVucHNXV05FUWtKV1dHeFJVVzFzY0ZWVmJHOVJWWGhZVkZod2VsVXhjRUpoVm1SWldraE9ObE13ZUZoUlZURTFZakF4VEVOc1FrNWphbFpQVjI1a1ZFMHhVVFJUYlVVd1drWm9WMkpWU2xkVVZVWjJVakJPUkdOVlpGUlVWRkUxVVd0R1RsSkZSWGxoTUVaT1VqRnNSRlJXUmtST1IxVXhZMnQwUjJOWFVraGlSVnByVkhwS1JXVnRhMHRTTTJzellXcGFTbFpyY0Zwa1JVNW9VVk4wY21GcGRGVk1NRTEzVmxkb2Qxb3dlRTlPTW1neVpFZFdXRTlZWkd0VFZYZHlWVEZzTUdKcVVuWldSemxFVkZaR1JFMXRjM2RPUkZaR1pHdDBRbU5VWkVoU2QzQkZZbTVuZDJGRmFISlVWMHB0VXpKYVMwc3dUa05hYWxZeFltcGFVMDlFYTNoUFYzQlRWREZhVm1SVlpFVmxWV3hNVlZkb1JsTnBPVkpSYWtKb1lWWldNMkpFWnpsRGFUQjBURk13ZEZKVk5VVkpSVTVHVld4U1NsSnJiRVJSVmxKR1RGTXdkRXhUTUVzaWZYMTlmUT09IiwiaW50ZWdyYXRlZFRpbWUiOjE2NzQyMzM2NDIsImxvZ0lEIjoiZDMyZjMwYTNjMzJkNjM5YzJiNzYyMjA1YTIxYzdiYjA3Nzg4ZTY4MjgzYTRhZTZmNDIxMTg3MjNhMWJlYTQ5NiIsImxvZ0luZGV4IjoyMjQyOTY5fQ=="
            }
        ]
    },
    "messageSignature": {
        "messageDigest": {
            "algorithm": "SHA2_256",
            "digest": "Ib+a4NLZrTt/tk8tYb0hIbHHXusYswz3L68RAv0q/Pc="
        },
        "signature": "MGYCMQCGth+2fxxF5z5yGE9dqdQzWAc6WFuPpVCZHr03Cdwk43egPV5D8XNV3ngOoF4FWYoCMQCL1u5RaqL/D2GIxxcBlcDp0Ktu+a09iUKm0QAGuPfJ0F1zHuCiRRX/3v0+X5SsiIU="
    }
}
```

That bundle signs for `README.md` as of [a0377a4](https://github.com/sigstore/sigstore-python/pull/465/commits/a0377a4f4bfc04c96e49ed1950826b27cfdaca5e).

**Comment by woodruffw** [comment]
cc @znewman01 and @kommendorkapten for visibility 🙂

File: sigstore/sign.py
Comment (woodruffw) [inline sigstore/sign.py:244]: cc @kommendorkapten do you have opinions on this? `kind_version` isn't marked as mandatory and the information it provides is probably duplicated in the log entry body itself, so I wasn't sure if it makes sense to include it here.

File: sigstore/sign.py
Comment (woodruffw) [inline sigstore/sign.py:253]: Flagging for review: I _think_ this is right, but I'm not 100% sure -- our `encode_canonical` produces a bytestring of the canonicalized JSON for the log entry, which _should_ be the same thing that Rekor's REST API returns.

File: sigstore/sign.py
Comment (woodruffw) [inline sigstore/sign.py:225]: Flagging for review: I wasn't sure about this either. During discussions in protobuf-specs it sounds like the root cert isn't expected to be included, but that clients *may* include any intermediates in use.

**Comment by woodruffw** [comment]
Currently evaluating against `sigstore-js`; blocked on resolution of https://github.com/sigstore/protobuf-specs/issues/62

File: sigstore/sign.py
Comment (woodruffw) [reply sigstore/sign.py]: `sigstore-js` doesn't seem to have problems verifying even without us embedding the intermediate, so I think this is probably fine for now.

File: sigstore/sign.py
Comment (woodruffw) [inline sigstore/sign.py:216]: This can wait until a follow-up, but the amount of round-tripping here (`str -> bytes`, base64 decoding, etc) suggests that the current `SigningResult` field types aren't ideal 🙂 

They're already part of the public API, however, so any internal changes to them will probably need backwards-compatible properties.

File: sigstore/_cli.py
Comment (znewman01) [inline sigstore/_cli.py:352]: My instinct would be to make this take a path

(But totally up to you!)

File: sigstore/sign.py
Comment (znewman01) [reply sigstore/sign.py:216]: +1 to the sentiment, but I understand the need for compatibility :)

File: sigstore/sign.py
Comment (znewman01) [inline sigstore/sign.py:222]: I think so

https://github.com/sigstore/protobuf-specs/blob/16541696de137c6281d66d075a4924d9bbd181ff/protos/sigstore_common.proto#L170-L173

File: sigstore/sign.py
Comment (znewman01) [reply sigstore/sign.py:244]: See also https://github.com/sigstore/protobuf-specs/issues/52

If I'm reading correctly we should keep it here because it may disappear from the log entry later.

**Review by znewman01 — DISMISSED** [review]
Overall looks great

File: sigstore/sign.py
Comment (znewman01) [reply sigstore/sign.py:222]: Oh, resolve this comment we can follow up in the thread you opened already

File: sigstore/sign.py
Comment (znewman01) [reply sigstore/sign.py]: I think you're supposed to include it based on the docs?

https://github.com/sigstore/protobuf-specs/blob/16541696de137c6281d66d075a4924d9bbd181ff/protos/sigstore_common.proto#L170-L173

CC @haydentherapper  who will have a strong opinion here

Comment (woodruffw) [reply sigstore/sign.py]: Hmm, my reading of the language there:

```
        // The first certificate in the array must be the leaf
        // certificate used for signing. Any intermediate certificates
        // must be stored as offset 1 to n-1, and the root certificate at
        // position n.
```

Is that the certificate list MUST be nonempty, the first member MUST be the leaf, and *any if present* intermediates and root MUST be listed in chain order. In other words, just the leaf is fine, but @haydentherapper can confirm!

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py:352]: Yeah, I need to think a bit about this -- `sigstore sign` can take multiple inputs, so `sigstore sign foo.txt bar.txt --bundle foo.txt` doesn't fit quite right :slightly_smiling_face: 

I'll probably do a variant of what I do for the other output flags: `--bundle` will actually enable the bundle output format, and then maybe `--sigstore-bundle <foo>` will actually write to `<fOo>` if exactly one input is provided. But that isn't great for UX either :upside_down_face:

File: sigstore/sign.py
Comment (woodruffw) [reply sigstore/sign.py:244]: Makes sense! In that case, am I right in thinking that `kind_version` should probably be marked as mandatory in the protobuf definitions?

File: sigstore/sign.py
Comment (woodruffw) [inline sigstore/sign.py:261]: It'd be nice if there was some way to default to this `media_type`, but I have no idea if that's possible :slightly_smiling_face:

File: sigstore/_cli.py
Comment (znewman01) [reply sigstore/_cli.py:352]: Ahh interesting...yeah, hard to find anything that makes sense here.

I wonder whether you need two verbs, a `sign-one` and a `sign-many`.

Happy to let it stand for now

File: sigstore/sign.py
Comment (znewman01) [reply sigstore/sign.py:261]: IIUC no :(

That said, I think that's sort-of by design? You should only set that `media_type` if you immediately plan to serialize to JSON

File: sigstore/sign.py
Comment (znewman01) [reply sigstore/sign.py]: Yeah, we could stand to have an additional clarifying comment for `verification_material`.

File: sigstore/sign.py
Comment (woodruffw) [reply sigstore/sign.py:261]: Aha, I missed that detail! That makes sense, then, and it's not a huge deal here.

File: sigstore/sign.py
Comment (kommendorkapten) [reply sigstore/sign.py:244]: Thanks for the link @znewman01 !

Yes, that should be mandatory in the spec. It's being added as of this: https://github.com/sigstore/protobuf-specs/pull/64/files

File: sigstore/sign.py
Comment (kommendorkapten) [reply sigstore/sign.py]: The intent was that the leaf certificate MUST be included. Any intermediate (if existing), the and the root are RECOMMENDED to be included. 

Regardless if they are included or not in the bundle. The intermediates and the root must of course be verified against the trust root, as this is the only way to securely verify that they are trusted. The reason for not only depend on the root certificate, but also the intermediate is that we don't ship any CRLs, if an intermediate is revoed, it's removed from the TUF root, and so a client MUST NOT trust it.

Comment (kommendorkapten) [reply sigstore/sign.py]: But agree, lets clarify this. The validation part is covered by the client spec, but I think the language is a bit unclear there too.

File: sigstore/sign.py
Comment (kommendorkapten) [reply sigstore/sign.py:261]: Yeah, it would be nice to have const values in the language bindings that would capture the current media type used.

Comment (woodruffw) [reply sigstore/sign.py:261]: 100% agreed -- I'm not sure how best to accomplish that without hacky patches to the codegen, though...

File: sigstore/sign.py
Comment (haydentherapper) [reply sigstore/sign.py]: Historical context from Web PKI: Root CAs that follow CA/B forum requirements (typically those that are a part of OS or Mozilla trust stores) have their signing keys offline, and have long lifetimes (10-20 years). Intermediate CAs are created with online signing keys and shorter lifetimes (~3 years) to issue end-entity (leaf) certificates. Because these signing keys are online, there's a higher chance of compromise. In the event of a compromise, a root CA maintains a CRL containing intermediate CA certs that clients are expected to check for intermediate CA trustworthiness.

TLS connections provide the full cert chain, because intermediates are meant only for chain building, to find a path from an end-entity certificate to a root. The root is provided I assume just for auditing purposes, but a verifier should never trust a root provided in a chain, it should come from a trust store. The intermediates build a chain from the leaf to the root, with each intermediate's status being checked against a CRL.

Back to the present: The Sigstore public good instance's root CA is online, because maintaining offline CAs is a pain, especially with this model of cross-company production ownership. We could have skipped using an intermediate and issued end-entity certs directly off of the root, but we chose to leverage an intermediate because it allowed us to convert the Sigstore root CA into a PKI root for all use-cases across Sigstore (currently just Fulcio, but it could include a timestamp authority too).

We didn't want to maintain CRLs because clients struggle with checking them. Either you need to always query an CRL online (no more offline verification then), or cache CRLs and pick some cadence to update. Size can also become an issue, but not worried about that here since root CA CRLs rarely grow large.

But we don't need CRLs for the public good instance, because we have TUF! We put the intermediate and the root in TUF, and so if either become untrusted, we simply remove them from the TUF trust root. Clients (Cosign for example) fetch the verification material from TUF, and for OCI, ignore any certs set in the `chain` annotation (if the TUF root does not contain intermediates, then Cosign does read only the intermediates, not the root, from the chain annotation)

**tl;dr - For cases of verification from the public good instance, it is sufficient to include only the leaf/end-entity certificate in the bundle**

But wait! But what about non-public good instance Sigstore, private Sigstore instances? If they aren't using TUF, this is an issue if they want to rely on a more standard-pki-like experience. Maybe they don't want to ship intermediates in their trust root and rely on CRLs (there was an issue in Cosign about using CRLs). For these use-cases, it would be preferable to include the full chain in the bundle (the root is never necessary, but like I said above, I think it's useful for auditing purposes).

So **second tl;dr - I would consider including the full chain, but if you don't want to support this now, I would at least document this as a limitation/isssue**

Comment (woodruffw) [reply sigstore/sign.py]: Thanks for the extremely detailed explanation @haydentherapper! 

Including the intermediate and root certs would require some modifications to the current `SigningResult` model, but not impractical ones. Given that, I'll add a comment here and do that in a follow-up (since I plan on more generally refactoring `SigningResult` per https://github.com/sigstore/sigstore-python/pull/465#discussion_r1082935187 anyways).

Comment (woodruffw) [reply sigstore/sign.py]: 56f0636..8651f95

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py:352]: Hmm, here's another option: instead of allowing users to opt into Sigstore bundles with `--bundle`, maybe we enable it by default and require them to disable it with `--no-bundle`? This would come at the cost of slightly confusing files created by default (we'd essentially dupe the verification materials on disk), but we'd also end up getting the format upgrade for "free" since users wouldn't have to remember to update their CLI. 

`--bundle` itself could then be `--bundle OUTPUT`, similar to the existing options for that.

Thoughts @di?

Comment (woodruffw) [reply sigstore/_cli.py:352]: (This would also maybe make deprecating the existing outputs a little easier: we could begin warning that `{input}.{crt,sig,rekor}` are deprecated, and then hide them behind a `--legacy-outputs` or similar flag in an upcoming major release.

File: sigstore/sign.py
Comment (kommendorkapten) [reply sigstore/sign.py:259]: Yes, would be messy. I would love for Protobufs to support that natively, so many use-cases I have seen where it would make our lives easier.

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py:352]: From discussion with @tetsuo-cpp: I'm convinced that `--no-bundle` is a better route here, and that we should emit the bundle by default. From there, `--bundle` can have the same behavior as the other single-output options do (i.e. write to the given file, rather than `{input}.bundle`).

Comment (tetsuo-cpp) [reply sigstore/_cli.py:352]: I have a commit that does this on the way.

File: sigstore/_cli.py
Comment (tetsuo-cpp) [inline sigstore/_cli.py:742]: I changed these back from:
```python
if "rekor_bundle" in outputs:
```
This is because when we don't write to an output, the entry still gets added to the map, the file is just set to `None`.

**Review by tetsuo-cpp — DISMISSED** [review]
LGTM!

@woodruffw I pushed a few commits to make the CLI work as we described, so please take a look at that when you get a moment. Other than that, this looks great. 🎉

File: sigstore/_cli.py
Comment (di) [reply sigstore/_cli.py:353]: I agree with the plan to use `--no-bundle` and have bundles generated by default.

File: sigstore/_cli.py
Comment (di) [inline sigstore/_cli.py:728]: This is a bug: the value `outputs["sig"]` will never be `in outputs` (the set of keys in `outputs`), and thus it will never write the `.sig` file.

I think this was meant to be:

```suggestion
        if outputs["sig"] is not None:
```

(I'm also curious why this wasn't caught by a test.)

**Review by di — CHANGES_REQUESTED** [review]
Overall LGTM, aside from the bug with writing `.sig` files.

Just a note: we probably don't want to make a release with this until verification is supported as well.

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py:728]: Yeah, this is supposed to be `"sig" in outputs` since we don't actually know the `"sig"` key is even present; this looks like a bad bulk edit on my part.

(I think we didn't catch this because the CLI code still isn't well-covered. But this is a good incentive to begin to raise it 🙂)

Comment (woodruffw) [reply sigstore/_cli.py]: d045d21

**Comment by woodruffw** [comment]
> Just a note: we probably don't want to make a release with this until verification is supported as well.

Sounds good!

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py:742]: Makes sense! I've re-fixed https://github.com/sigstore/sigstore-python/pull/465#discussion_r1086729138 based on that.

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py]: (I looked at our tests again, and this would have been caught by the conformance suite once it merged on main -- this is making me think that we should probably also run the conformance suite on PRs.)

**Review by di — APPROVED** [review]

**Pull request merged**
